e2e/push/push.go

```diff
@@ -75,26 +76,23 @@ func (c *ctx) testPushCmd(t *testing.T) {
 			}
 			defer os.RemoveAll(tmpdir)
 
-			cmd, out, err := e2e.ImagePush(t, c.env.CmdPath, tt.imagePath, tt.dstURI)
-			switch {
-			case tt.expectSuccess && err == nil:
-				// PASS: expecting success, succeeded
-
-			case !tt.expectSuccess && err != nil:
-				// PASS: expecting failure, failed
-
-			case tt.expectSuccess && err != nil:
-				// FAIL: expecting success, failed
-
-				t.Logf("Running command:\n%s\nOutput:\n%s\n", cmd, out)
-				t.Errorf("unexpected failure: %v", err)
-
-			case !tt.expectSuccess && err == nil:
-				// FAIL: expecting failure, succeeded
-
-				t.Logf("Running command:\n%s\nOutput:\n%s\n", cmd, out)
-				t.Errorf("unexpected success: command should have failed")
+			// We create the list of arguments using a string instead of a slice of
+			// strings because using slices of strings most of the type ends up adding
+			// an empty elements to the list when passing it to the command, which
+			// will create a failure.
+			args := tt.dstURI
+			if tt.imagePath != "" {
+				args = tt.imagePath + " " + args
 			}
+
+			e2e.RunSingularity(
+				t,
+				tt.desc,
```

Comment:
replace this with `e2e.AsSubtest(tt.desc)`
```suggestion
				e2e.AsSubtest(tt.desc),
```

Reply:
That is the part I do not understand, semantically, I guess I do not understand the meaning of `AsSubtest`, which, without spending more than 2 minutes on it, would infer that I want to create a subtest... which is not the case. But I will look it up, I just need time. :)

Reply:
Welcome to my world :-)

The function `e2e.RunSingularity` previously had a signature that was something like `func(t *testing.T, name string, ...)`.

I changed it to `func(t *testing.T, cmdPath string, ...)`

In the previous signature, `name` is the name of the subtest (as in `t.Run(name, f)`). If name was empty (`""`), then it _did not_ run a subtest. If name was not empty, it run the command as a subtest.

To make that less awkward and less magic, I changed it to the explicit `e2e.AsSubtest(name)`, which is passed along with all the other options.

But then I added the `cmdPath` argument :-(

The `cmdPath` argument specifies the path to singularity.

I guess I could add a `type CmdPath string` to make this a compiler error (at the cost of making the code a little weird).

Reply:
And if you don't want a subtest, just don't pass e2e.AsSubtest. I suggested that to match the behavior of the previous code.

Reply:
Thanks for the explanation, it truly helps. I like the fact that we can specify the Singularity binary and now that I have your explanation, I understand the subtest part too. I am totally okay with the new API. :)